subsys/bluetooth/lib/ead.c

```diff
@@ -0,0 +1,221 @@
+
+	if (err != 0) {
+		LOG_ERR("Failed to encrypt the payload (bt_ccm_encrypt err %d)", err);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+int bt_ead_encrypt(const uint8_t *session_key, const uint8_t *iv, const uint8_t *payload,
+		   size_t payload_size, uint8_t *encrypted_payload)
+{
+	if (session_key == NULL || iv == NULL || payload == NULL || encrypted_payload == NULL) {
+		LOG_ERR("At least one of the argument is a NULL pointer.");
+		return -EINVAL;
+	}
+
+	if (payload_size == 0) {
+		LOG_WRN("payload_size is set to 0. The encrypted result will be filled with only "
+			"the Randomizer and the MIC.");
+	}
+
+	return ead_encrypt(session_key, iv, NULL, payload, payload_size, encrypted_payload);
+}
+
+#ifdef CONFIG_BT_TESTING
+
+int bt_test_ead_encrypt(const uint8_t *session_key, const uint8_t *iv, const uint8_t *randomizer,
+			const uint8_t *payload, size_t payload_size, uint8_t *encrypted_payload)
+{
+	if (session_key == NULL || iv == NULL || payload == NULL || encrypted_payload == NULL) {
+		LOG_ERR("At least one of the argument is a NULL pointer.");
+		return -EINVAL;
+	}
+
+	if (payload_size == 0) {
+		LOG_WRN("payload_size is set to 0. The encrypted result will be filled with only "
+			"the Randomizer and the MIC.");
+	}
+
+	return ead_encrypt(session_key, iv, randomizer, payload, payload_size, encrypted_payload);
+}
+
+#endif
+
+int ead_decrypt(const uint8_t *session_key, const uint8_t *iv, const uint8_t *encrypted_payload,
+		size_t encrypted_payload_size, uint8_t *payload)
+{
+	int err;
+	uint8_t nonce[BT_EAD_NONCE_SIZE];
+	const uint8_t *encrypted_ad_data = &encrypted_payload[BT_EAD_RANDOMIZER_SIZE];
+	size_t encrypted_ad_data_size = encrypted_payload_size - BT_EAD_RANDOMIZER_SIZE;
+	size_t payload_size = encrypted_ad_data_size - BT_EAD_MIC_SIZE;
+
+	const uint8_t *randomizer = encrypted_payload;
+
+	err = bt_ead_generate_nonce(iv, randomizer, nonce);
+
+	if (err != 0) {
+		return -EIO;
+	}
+
+	LOG_HEXDUMP_DBG(encrypted_ad_data, encrypted_ad_data_size, "Encrypted Data: ");
+
+	err = bt_ccm_decrypt(session_key, nonce, encrypted_ad_data, payload_size, bt_ead_aad,
+			     BT_EAD_AAD_SIZE, payload, BT_EAD_MIC_SIZE);
+
+	if (err != 0) {
+		LOG_ERR("Failed to decrypt the data (bt_ccm_decrypt err %d)", err);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+int bt_ead_decrypt(const uint8_t *session_key, const uint8_t *iv, const uint8_t *encrypted_payload,
+		   size_t encrypted_payload_size, uint8_t *payload)
+{
+	if (session_key == NULL || iv == NULL || encrypted_payload == NULL || payload == NULL) {
+		LOG_ERR("At least one of the argument is a NULL pointer.");
+		return -EINVAL;
+	}
+
+	if (encrypted_payload_size < 0) {
+		LOG_ERR("encrypted_payload_size is set to %d. The minimum size of "
+			"encrypted_payload_size is BT_EAD_RANDOMIZER_SIZE + BT_EAD_MIC_SIZE (%d).",
+			encrypted_payload_size, BT_EAD_RANDOMIZER_SIZE + BT_EAD_MIC_SIZE);
+		return -EINVAL;
+	} else if (encrypted_payload_size == BT_EAD_RANDOMIZER_SIZE + BT_EAD_MIC_SIZE) {
```

Comment:
Can we just do `encrypted_payload_size < BT_EAD_RANDOMIZER_SIZE + BT_EAD_MIC_SIZE`?

Reply:
Yes, you're right, I updated it. But yeah, I the second one is useful for the warning.